### PR TITLE
(#30) Add pre-commit with ruff + black

### PR DESCRIPTION
**Pull request by coffeewasmyidea**
Hello Stanislav.

I saw this issue on your list, and here it is a small contribution from my side. It's nothing super special, but still. 

I hope this helps. 

With the best regards, Sergey.

### CODE REVIEW

**Comment by coffeewasmyidea** [comment]
If you want to use Pyright in pre-commit also, just take this `.pre-commit-config.yaml`:

```shell
default_language_version:
  python: python3.10
repos:
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v4.4.0
    hooks:
    - id: check-added-large-files
    - id: check-yaml
    - id: check-toml
    - id: end-of-file-fixer
    - id: trailing-whitespace
    - id: detect-private-key

  - repo: https://github.com/pre-commit/pygrep-hooks
    rev: v1.10.0
    hooks:
      - id: python-check-blanket-noqa

  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.0.284
    hooks:
      - id: ruff
        args: [--fix, --exit-non-zero-on-fix]

  - repo: https://github.com/psf/black
    rev: 23.7.0
    hooks:
    - id: black

  - repo: https://github.com/PyCQA/bandit
    rev: 1.7.5
    hooks:
    - id: bandit
      args: ["--skip=B101,B311", "--exclude=tests"]

  - repo: https://github.com/RobertCraigie/pyright-python
    rev: v1.1.322
    hooks:
    - id: pyright
```

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    hooks:
+      - id: ruff
+        args: [--fix, --exit-non-zero-on-fix]
+
+  - repo: https://github.com/psf/black
+    rev: 23.7.0
+    hooks:
+    - id: black
+
+  - repo: https://github.com/PyCQA/bandit
```
Comment (Ovsyanka83) [inline .pre-commit-config.yaml:30]: No need too have bandit: ruff includes it.

**Review by Ovsyanka83 — CHANGES_REQUESTED** [review]
Thanks for the PR!

I feel like it would make sense to revert changes in *.py because all fixes in *.py files are either breaking the existing behavior (changing unions), using less effective checks (unions instead of tuples in isinstance), or removing noqa comments from places where I would prefer for them to stay (specifically the complexity noqas because I rely on them as a reminder to improve complexity some time later).

I also left a few more comments. Hope you have time to take a look.

File: foo.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
```
[the remark this answers is not in the document]
Comment (coffeewasmyidea) [reply foo.json:1]: Hmm, it is not me. this is a fix from `end-of-file-fixer`.

File: pyproject.toml [hunk 3]
```diff
@@ -142,6 +152,22 @@ ignore = [
     "RET505", # Unnecessary `else` after `return` statement
     "N805",   # First argument of a method should be named `self`
 ]
+exclude = [
```
[the remark this answers is not in the document]
Comment (coffeewasmyidea) [reply pyproject.toml:155]: Yes, it does, but in this section, you can add something specific that needs to be excluded from linting. This is from the documentation. Or do you think that section will also should be removed?

File: pyproject.toml [hunk 4]
```diff
@@ -155,6 +181,12 @@ extend-exclude = ["migrations"]
     "ERA001", # Found commented-out code (it's not actually commented out. It's just comments)
 ]
 
+[tool.ruff.mccabe]
+max-complexity = 14
+
+[tool.bandit]
```
[the remark this answers is not in the document]
Comment (coffeewasmyidea) [reply pyproject.toml:187]: okay, removed.

File: pyproject.toml [hunk 5]
```diff
@@ -155,6 +181,12 @@ extend-exclude = ["migrations"]
     "ERA001", # Found commented-out code (it's not actually commented out. It's just comments)
 ]
 
+[tool.ruff.mccabe]
+max-complexity = 14
```
[the remark this answers is not in the document]
Comment (coffeewasmyidea) [reply pyproject.toml:185]: okay, removed.

File: universi/_utils.py [hunk 6]
```diff
@@ -2,14 +2,15 @@
...
+from collections.abc import Collection
 from pathlib import Path
 from types import ModuleType
-from typing import Any, Collection, Union
+from typing import Any
 
 from universi.exceptions import UniversiError
 
 Sentinel: Any = object()
-UnionType = type(int | str) | type(Union[int, str])
```
[the remark this answers is not in the document]
Comment (coffeewasmyidea) [reply universi/_utils.py:12]: I also not me :) okay, I reverted :ok_hand:

File: pyproject.toml [hunk 7]
```diff
@@ -79,10 +80,19 @@ exclude_lines = [
 ]
 omit = ["tests/test_tutorial/test_users_example003/*"]
 
+[tool.isort]
```
[the remark this answers is not in the document]
Comment (coffeewasmyidea) [reply pyproject.toml:83]: If you don't plan to use isort, then you don't need it, you're right.

File: tests/test_codegen.py [hunk 8]
```diff
@@ -496,7 +496,7 @@ def test__codegen__with_weird_data_types(generate_test_version_packages: Generat
 def test__codegen_union_fields(generate_test_version_packages: GenerateTestVersionPackages):
     v2000_01_01, v2001_01_01 = generate_test_version_packages(
         schema(latest.SchemaWithUnionFields).field("baz").existed_with(type=int | latest.EmptySchema),
-        schema(latest.SchemaWithUnionFields).field("daz").existed_with(type=Union[int, latest.EmptySchema]),
+        schema(latest.SchemaWithUnionFields).field("daz").existed_with(type=int | latest.EmptySchema),
```
[the remark this answers is not in the document]
Comment (coffeewasmyidea) [reply tests/test_codegen.py:499]: I also not me :) okay, I reverted 👌

File: .pre-commit-config.yaml (hunk 1 above)
Comment (coffeewasmyidea) [reply .pre-commit-config.yaml:30]: Okay, I reverted 👌

File: pyproject.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ovsyanka83) [reply pyproject.toml:155]: What you've added are the **defaults**.  I.e. What ruff already does :)

[the remark this answers is not in the document]
Comment (coffeewasmyidea) [reply pyproject.toml]: okay, removed.

**Comment by coffeewasmyidea** [comment]
It seems that everything that needs to be done has been done. But keep in mind that half of my reverse was auto-fixed by the pre-commit configuration, so you need to pay attention to this to prevent double work.

P.s. As a good practice, we kept only the complexity equal to 14.

**Review by Ovsyanka83 — APPROVED** [review]
Thanks for your work!

**Pull request merged**